In [1]:
%cd ..

D:\SoftUni\Machine Learning\mushrooms project


In [2]:
import pandas as pd
pd.set_option('display.max_columns', None)
import matplotlib.pyplot as plt
import seaborn as sns

import numpy as np

from sklearn.model_selection import train_test_split, StratifiedKFold, cross_val_score
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import OneHotEncoder
from sklearn.linear_model import LogisticRegression
from sklearn.naive_bayes import CategoricalNB
from sklearn.ensemble import RandomForestClassifier
from sklearn.svm import SVC
from sklearn.metrics import make_scorer, accuracy_score, recall_score, roc_auc_score 
from sklearn.model_selection import cross_validate

from hyperopt import fmin, tpe, hp, Trials, STATUS_OK

from src import functions

In [3]:
mushroom_data = pd.read_csv('data/cleaned_dataset.csv')

In [4]:
mushroom_data

label cap_shape cap_surface cap_color bruises odor gill_attachment  \
0        p         x           s         n       t    p               f   
1        e         x           s         y       t    a               f   
2        e         b           s         w       t    l               f   
3        p         x           y         w       t    p               f   
4        e         x           s         g       f    n               f   
...    ...       ...         ...       ...     ...  ...             ...   
8119     e         k           s         n       f    n               a   
8120     e         x           s         n       f    n               a   
8121     e         f           s         n       f    n               a   
8122     p         k           y         n       f    y               f   
8123     e         x           s         n       f    n               a   

     gill_spacing gill_size gill_color stalk_shape stalk_root  \
0               c         n          k           e          e   
1               c         b          k           e          c   
2               c         b          n           e          c   
3               c         n          n           e          e   
4               w         b          k           t          e   
...           ...       ...        ...         ...        ...   
8119            c         b          y           e          ?   
8120            c         b          y           e          ?   
8121            c         b          n           e          ?   
8122            c         n          b           t          ?   
8123            c         b          y           e          ?   

     stalk_surface_above_ring stalk_surface_below_ring stalk_color_above_ring  \
0                           s                        s                      w   
1                           s                        s                      w   
2                           s                        s                      w   
3                           s                        s                      w   
4                           s                        s                      w   
...                       ...                      ...                    ...   
8119                        s                        s                      o   
8120                        s                        s                      o   
8121                        s                        s                      o   
8122                        s                        k                      w   
8123                        s                        s                      o   

     stalk_color_below_ring veil_color ring_number ring_type  \
0                         w          w           o         p   
1                         w          w           o         p   
2                         w          w           o         p   
3                         w          w           o         p   
4                         w          w           o         e   
...                     ...        ...         ...       ...   
8119                      o          o           o         p   
8120                      o          n           o         p   
8121                      o          o           o         p   
8122                      w          w           o         e   
8123                      o          o           o         p   

     spore_print_color population habitat  
0                    k          s       u  
1                    n          n       g  
2                    n          n       m  
3                    k          s       u  
4                    n          a       g  
...                ...        ...     ...  
8119                 b          c       l  
8120                 b          v       l  
8121                 b          c       l  
8122                 w          v       l  
8123                 o          c       l  

[8124 rows x 22 columns]

### Spliting the dataset
Sklearn "train_test_split" method is used to split the data to train and test subsets. It is shuffled before the split and is stratified by the target variable to preserve it's classes proportion. Accordingly to the size of the dataset and the presence of some rare categories, a 75 / 25 split us used to ensure a decent size and quality of the testing set.

In [5]:
RANDOM_SEED = 11

In [6]:
mushroom_attributes = mushroom_data.drop(columns = ['label'])

In [7]:
mushroom_target = mushroom_data['label']

In [8]:
mushroom_attr_train, mushroom_attr_test, mushroom_target_train, mushroom_target_test = train_test_split(
    mushroom_attributes, mushroom_target, test_size = 0.25, random_state = RANDOM_SEED, stratify = mushroom_target
)

In [9]:
mushroom_attr_train.shape, mushroom_attr_test.shape, mushroom_target_train.shape, mushroom_target_test.shape

((6093, 21), (2031, 21), (6093,), (2031,))

The proportions of target variable have remained the same after the split:

In [10]:
mushroom_data['label'].value_counts(normalize = True)

label
e    0.517971
p    0.482029
Name: proportion, dtype: float64

In [11]:
mushroom_target_train.value_counts(normalize = True)

label
e    0.517971
p    0.482029
Name: proportion, dtype: float64

In [12]:
mushroom_target_test.value_counts(normalize = True)

label
e    0.517971
p    0.482029
Name: proportion, dtype: float64

### Encoding
All feature's values have to be encoded from "str" to numerical data type as needed for proper work with ML algorithms. I will use one - hot encoding since the categories within the features are not ranked or hierarchical, thus label encoding would not be appropriate. Because the presence of some rare categhories, handle_unknown = "ignore" and a list which contains all possible categories are provided as arguments to the encoder.

In [13]:
mushroom_columns = mushroom_attr_train.columns.values

In [14]:
mushroom_categories = [mushroom_attr_train[col].unique() for col in mushroom_columns]

In [15]:
attr_encoder = ColumnTransformer(
    transformers = [
        ('one-hot', OneHotEncoder(handle_unknown = 'ignore', categories = mushroom_categories), mushroom_columns)
    ]
)

### Algorithms

Algorithm selection is based on the applied EDA, characteristics of the dataset and results of previous research. Even with simple algorithms, high values of evaluation metrics have been achieved previously. Indeed, the results of the data analysis showed that some of the features can be of great importance in determining the class of the target variable. At the other hand, the size of the dataset may be not big enough to complex algorithms like XGBoost which are prone to overfitting. Simple classification algorithm like Logistic Regression which can be used initially as a baseline. Applying Random Forest and SVC could rise the performance(w.r.t the evaluation metric) afterwards if needed. Coefficients of Logistic Regression and RF feature importances are usable  

### Evaluation and hyperparameter tuning
To assess the performance of the models is used stratified k - fold cross - validation technique. This approach is suitable because the dataset is relatively small and do not have enough obervations for separate validation set. For this reasos the training set is randomly divided into 5 folds, ensuring that every data point, including those with rare categories, is used in both training and validation.

To distinguish the toxic mushrooms is of great importance. Labeling a poisonous mushroom as edible can be fatal in the real world. For this reasont I want to ensure that when the model predicts "edible" it is actually edible. The appropriate evaluation metric in this case is **recall**:

$$ Recall = \frac{TP}{TP + FN} $$

It shows what part of all "poisonous" labeled mushrooms was retrieved by the model. I will use it as main evaluation metric to asses the performance of the classification models in the project.

Hyperopt library is used for hyperparameter tuning. Negative recall is set as loss metric for optimization since the algorithm tries to minimize it. In addition, accuracy and area under the ROC metrics are observed. Appropriate search space and objective function is defined for each of the used algorithms.

In [16]:
eval_metrics = {
        'recall': make_scorer(recall_score, pos_label = 'p'),
        'accuracy': make_scorer(accuracy_score),
        'roc_auc': 'roc_auc'
    }

#### Logarithmic Regression

In [17]:
# def logr_objective(params):
#     pipeline = Pipeline(steps = [
#         ('encode', attr_encoder),
#         ('classifier', LogisticRegression(solver = 'liblinear', random_state = RANDOM_SEED))
#     ])

#     pipeline.set_params(
#         classifier__C = params['C'],
#         classifier__penalty = params['penalty']
#     )

#     skf = StratifiedKFold(n_splits = 5)
    
#     scores = cross_validate(pipeline, mushroom_attr_train, mushroom_target_train, cv = skf, scoring = eval_metrics)

#     return {
#         'loss': -np.mean(scores['test_recall']),
#         'status': STATUS_OK,
#         'other_metrics': {
#             'mean_recall': np.mean(scores['test_recall']),
#             'mean_accuracy': np.mean(scores['test_accuracy']),
#             'mean_roc_auc': np.mean(scores['test_roc_auc'])
#         }
#     }

In [20]:
logr_search_space = {
    'penalty': hp.choice('penalty', ['l1', 'l2']),
    'C': hp.loguniform('C', np.log(1e-3), np.log(1e2))
}

In [21]:
logr_trials = Trials()

logr_fmin = fmin(
    fn = lambda params: functions.hyperopt_objective(
        params = params,
        classifier = 'LogisticRegression', 
        attr_train = mushroom_attr_train,
        target_train = mushroom_target_train,
        eval_metrics = eval_metrics,
        encoder = attr_encoder,
        random_seed = RANDOM_SEED
    ),
    space = logr_search_space,
    algo = tpe.suggest,
    max_evals = 80,
    trials = logr_trials,
    rstate = np.random.default_rng(RANDOM_SEED)
)

100%|██████████████████████████████████████████████████████████████| 80/80 [00:49<00:00,  1.61trial/s, best loss: -1.0]


In [22]:
logr_fmin

{'C': 68.29736138744246, 'penalty': 1}

Minimization function returns "C" = 68.29736138744246 and 'penalty' = 1(L2) as the best values of hyperparameters.

The relevant information from each trial is extracted and visualized as pandas dataframe.

In [23]:
logr_metrics = []
for trial in logr_trials.trials:
    result = trial['result']
    logr_metrics.append({
        'loss': result['loss'],
        'mean_recall': result['other_metrics']['mean_recall'],
        'mean_accuracy': result['other_metrics']['mean_accuracy'],
        'mean_roc_auc': result['other_metrics']['mean_roc_auc'],
        'params': trial['misc']['vals']
    })

logr_trials_df = pd.DataFrame(logr_metrics)

In [24]:
logr_trials_df

loss  mean_recall  mean_accuracy  mean_roc_auc  \
0  -0.995574     0.995574       0.997866      0.999953   
1  -1.000000     1.000000       1.000000      1.000000   
2  -1.000000     1.000000       1.000000      1.000000   
3  -1.000000     1.000000       1.000000      1.000000   
4  -0.965613     0.965613       0.974396      0.996803   
..       ...          ...            ...           ...   
75 -0.997957     0.997957       0.999015      0.999989   
76 -1.000000     1.000000       1.000000      1.000000   
77 -1.000000     1.000000       1.000000      1.000000   
78 -1.000000     1.000000       1.000000      1.000000   
79 -1.000000     1.000000       1.000000      1.000000   

                                           params  
0    {'C': [0.09375928061459672], 'penalty': [1]}  
1      {'C': [68.29736138744246], 'penalty': [1]}  
2      {'C': [7.951969983018893], 'penalty': [0]}  
3      {'C': [4.220650205931255], 'penalty': [0]}  
4   {'C': [0.004064644237754485], 'penalty': [1]}  
..                                            ...  
75   {'C': [0.15577941981187796], 'penalty': [1]}  
76    {'C': [16.785735046250956], 'penalty': [1]}  
77     {'C': [23.29286159258232], 'penalty': [1]}  
78     {'C': [8.806770443443561], 'penalty': [1]}  
79     {'C': [2.340267050948972], 'penalty': [1]}  

[80 rows x 5 columns]

There are many combinations of hyperparameters which result in maximum value for the evaluation metrics. This may be sign of overfitting. To confirm whether it is occurring, the model will be evaluated on the test set. I suggest that the high values of the evaluation metrics ​​are due to the relatively small size of the dataset, balanced distribution of the target variable with easily separable classes, and presence of clear patterns in the data. 

From the trials with perfect score the combination of hyperparameters with the highest regularization strength among these using L1 penalty, is choosen. Since the results of EDA suggest that some features have low significance w.r.t the task of predicting the target variable label, Lasso is applied to sieve only the significant ones. Pushing the coefficients of the less important features to zero allows the importance of features to be interpreted. Since the modelling produces high scores of evaluation metrics, regularization strength is chosen to be of maximum value(among trials with L1 penalty) to reduce eventual overfitting.

In [25]:
logr_best_trial = logr_trials_df.sort_values(
    by = ['params'], 
    key = lambda col: col.apply(lambda x: (x['penalty'][0], x['C'][0]))
).sort_values(
    by = ['mean_recall', 'mean_accuracy', 'mean_roc_auc'], 
    ascending = False
).head(1)

In [26]:
logr_best_trial

loss  mean_recall  mean_accuracy  mean_roc_auc  \
59  -1.0          1.0            1.0           1.0   

                                         params  
59  {'C': [0.8859574659241594], 'penalty': [0]}

In [27]:
logr_c = logr_best_trial['params'].values[0]['C'][0]
logr_c

0.8859574659241594

In [28]:
logr_penalty = 'l1'

In [16]:
# def rf_objective(params):
#     pipeline = Pipeline(steps = [
#         ('encode', encoder),
#         ('classifier', RandomForestClassifier())
#     ])

#     pipeline.set_params(
#         classifier__n_estimators = int(params['n_estimators']),
#         classifier__max_depth = int(params['max_depth'])
#     )

#     skf = StratifiedKFold(n_splits = 5)
    
#     scores = cross_validate(pipeline, mushroom_attr_train, mushroom_target_train, cv = skf, scoring = eval_metrics)

#     return {
#         'loss': np.mean(scores['test_recall']),
#         'status': STATUS_OK,
#         'other_metrics': {
#             'mean_accuracy': np.mean(scores['test_accuracy']),
#             'mean_precision': np.mean(scores['test_precision']),
#             'mean_f1': np.mean(scores['test_f1']),
#         }
#     }

In [17]:
# rf_search_space = {
#     'n_estimators': hp.quniform('n_estimators', 20, 100, 20),
#     'max_depth': hp.quniform('max_depth', 5, 8, 1)
# }

In [18]:
# rf_trials = Trials()

# rf_best = fmin(
#     fn = lambda params: rf_objective(params),
#     space = rf_search_space,
#     algo = tpe.suggest,
#     max_evals = 40,
#     trials = rf_trials,
# )

100%|████████████████████████████████████████████████| 40/40 [01:28<00:00,  2.21s/trial, best loss: 0.9899886983731854]


In [19]:
# rf_best

{'max_depth': 5.0, 'n_estimators': 100.0}

In [20]:
# results_df = pd.DataFrame([
#     {
#         'iteration': i,
#         'loss': trial['result']['loss'],
#         'accuracy': trial['result']['other_metrics']['mean_accuracy'],
#         'precision': trial['result']['other_metrics']['mean_precision'],
#         'f1': trial['result']['other_metrics']['mean_f1'],
#         **trial['misc']['vals']
#     }
#     for i, trial in enumerate(rf_trials.trials)
# ])

In [42]:
# results_df

iteration      loss  accuracy  precision        f1 max_depth n_estimators
0           0  0.992286  0.992286   0.992406  0.992283     [5.0]       [60.0]
1           1  1.000000  1.000000   1.000000  1.000000     [7.0]       [60.0]
2           2  0.992286  0.992286   0.992414  0.992283     [5.0]       [20.0]
3           3  0.993435  0.993435   0.993542  0.993432     [6.0]       [60.0]
4           4  1.000000  1.000000   1.000000  1.000000     [8.0]       [40.0]
5           5  0.998687  0.998687   0.998692  0.998687     [6.0]       [20.0]
6           6  1.000000  1.000000   1.000000  1.000000     [8.0]       [20.0]
7           7  0.991301  0.991301   0.991454  0.991298     [5.0]       [80.0]
8           8  0.997867  0.997867   0.997886  0.997867     [6.0]       [80.0]
9           9  1.000000  1.000000   1.000000  1.000000     [8.0]      [100.0]
10         10  0.999672  0.999672   0.999672  0.999672     [7.0]       [40.0]
11         11  1.000000  1.000000   1.000000  1.000000     [7.0]       [60.0]
12         12  0.995568  0.995568   0.995647  0.995566     [6.0]       [80.0]
13         13  0.997867  0.997867   0.997886  0.997867     [6.0]      [100.0]
14         14  0.997538  0.997538   0.997554  0.997538     [6.0]       [80.0]
15         15  0.996717  0.996717   0.996746  0.996716     [6.0]       [60.0]
16         16  0.994748  0.994748   0.994803  0.994747     [6.0]       [40.0]
17         17  1.000000  1.000000   1.000000  1.000000     [7.0]      [100.0]
18         18  0.996717  0.996717   0.996763  0.996716     [6.0]       [40.0]
19         19  0.999836  0.999836   0.999836  0.999836     [7.0]       [80.0]
20         20  0.990153  0.990153   0.990343  0.990148     [5.0]       [80.0]
21         21  0.990317  0.990317   0.990504  0.990312     [5.0]       [80.0]
22         22  0.990809  0.990809   0.990981  0.990805     [5.0]      [100.0]
23         23  0.990317  0.990317   0.990502  0.990312     [5.0]       [80.0]
24         24  0.989989  0.989989   0.990185  0.989984     [5.0]      [100.0]
25         25  0.990973  0.990973   0.991137  0.990969     [5.0]      [100.0]
26         26  0.989989  0.989989   0.990185  0.989984     [5.0]      [100.0]
27         27  0.990317  0.990317   0.990501  0.990312     [5.0]      [100.0]
28         28  0.990481  0.990481   0.990668  0.990477     [5.0]      [100.0]
29         29  0.989989  0.989989   0.990185  0.989984     [5.0]      [100.0]
30         30  0.990809  0.990809   0.990977  0.990805     [5.0]      [100.0]
31         31  0.989989  0.989989   0.990185  0.989984     [5.0]      [100.0]
32         32  0.991466  0.991466   0.991618  0.991462     [5.0]       [60.0]
33         33  0.990974  0.990974   0.991138  0.990970     [5.0]       [80.0]
34         34  0.996553  0.996553   0.996586  0.996552     [6.0]      [100.0]
35         35  1.000000  1.000000   1.000000  1.000000     [8.0]       [60.0]
36         36  0.990152  0.990152   0.990306  0.990148     [5.0]       [20.0]
37         37  0.996389  0.996389   0.996437  0.996387     [6.0]       [80.0]
38         38  0.990153  0.990153   0.990343  0.990148     [5.0]      [100.0]
39         39  0.999672  0.999672   0.999673  0.999672     [7.0]       [40.0]

In [32]:
# best_params

{'LogisticRegression': {'C': 1.4333552786419543},
 'RandomForest': {'max_depth': 8.0, 'n_estimators': 70.0},
 'SVC': {'C': 2.890974943886087, 'kernel': 0}}